File: packages/core/api/endpoints/chat-threads.ts
import { z } from "zod";
import { request } from "../http";
import { parseWithFallback } from "../schema";
import {
  ChatMessageSchema,
  enc,
  type ChatMessageRecord,
} from "./chat-schemas";

const MessagesListSchema = z.object({
  messages: z.array(ChatMessageSchema).optional().default([]),
});

const MessageEnvelopeSchema = z.object({
  message: ChatMessageSchema.optional(),
});

const StatusSchema = z.object({
  ok: z.boolean().optional(),
});

const ChatThreadSchema = z.object({
  thread_root_id: z.string(),
  room_id: z.string(),
  workspace_id: z.string(),
  root_body: z.string().optional().default(""),
  root_sender_id: z.string().optional().default(""),
  reply_count: z.number().optional().default(0),
  last_reply_at: z.string().optional(),
  root_created_at: z.string().optional().default(""),
  unread: z.boolean().optional().default(false),
  reason: z.string().optional().default(""),
});

const ThreadListSchema = z.object({
  threads: z.array(ChatThreadSchema).optional().default([]),
});

export type ChatThreadRecord = z.infer<typeof ChatThreadSchema>;

export async function listChatThreadMessages(
  workspaceId: string,
  roomId: string,
  threadRootId: string,
  opts: { limit?: number; before?: string } = {},
): Promise<ChatMessageRecord[]> {
  const q = new URLSearchParams();
  if (opts.limit != null) q.set("limit", String(opts.limit));
  if (opts.before) q.set("before", opts.before);
  const qs = q.toString() ? `?${q}` : "";
  const raw = await request(
    `/api/v1/workspaces/${enc(workspaceId)}/chat/rooms/${enc(roomId)}/threads/${enc(threadRootId)}/messages${qs}`,
  );
  const parsed = parseWithFallback(raw, MessagesListSchema, { messages: [] }, {
    endpoint: "GET .../threads/{id}/messages",
  });
  return parsed.messages ?? [];
}

export async function sendChatThreadMessage(
  workspaceId: string,
  roomId: string,
  threadRootId: string,
  input: {
    body: string;
    client_msg_id?: string;
    priority?: string;
  },
): Promise<ChatMessageRecord | null> {
  const raw = await request(
    `/api/v1/workspaces/${enc(workspaceId)}/chat/rooms/${enc(roomId)}/threads/${enc(threadRootId)}/messages`,
    { method: "POST", body: input },
  );
  const parsed = parseWithFallback(raw, MessageEnvelopeSchema, { message: undefined }, {
    endpoint: "POST .../threads/{id}/messages",
  });
  return parsed.message ?? null;
}

export async function followChatThread(
  workspaceId: string,
  threadRootId: string,
): Promise<void> {
  const raw = await request(
    `/api/v1/workspaces/${enc(workspaceId)}/chat/threads/${enc(threadRootId)}/follow`,
    { method: "POST" },
  );
  parseWithFallback(raw, StatusSchema, { ok: true }, {
    endpoint: "POST .../threads/{id}/follow",
  });
}

export async function unfollowChatThread(
  workspaceId: string,
  threadRootId: string,
): Promise<void> {
  const raw = await request(
    `/api/v1/workspaces/${enc(workspaceId)}/chat/threads/${enc(threadRootId)}/follow`,
    { method: "DELETE" },
  );
  parseWithFallback(raw, StatusSchema, { ok: true }, {
    endpoint: "DELETE .../threads/{id}/follow",
  });
}

export async function markChatThreadRead(
  workspaceId: string,
  threadRootId: string,
): Promise<void> {
  const raw = await request(
    `/api/v1/workspaces/${enc(workspaceId)}/chat/threads/${enc(threadRootId)}/read`,
    { method: "POST" },
  );
  parseWithFallback(raw, StatusSchema, { ok: true }, {
    endpoint: "POST .../threads/{id}/read",
  });
}

export async function listFollowedChatThreads(
  workspaceId: string,
  opts: { unread?: boolean; limit?: number } = {},
): Promise<ChatThreadRecord[]> {
  const q = new URLSearchParams();
  if (opts.unread) q.set("unread", "1");
  if (opts.limit != null) q.set("limit", String(opts.limit));
  const qs = q.toString() ? `?${q}` : "";
  const raw = await request(
    `/api/v1/workspaces/${enc(workspaceId)}/chat/threads${qs}`,
  );
  const parsed = parseWithFallback(raw, ThreadListSchema, { threads: [] }, {
    endpoint: "GET .../chat/threads",
  });
  return parsed.threads ?? [];
}

File: server/internal/handler/chat_threads.go
package handler

import (
	"net/http"
	"strconv"
	"time"

	"github.com/go-chi/chi/v5"

	"github.com/unicomhub/uniwork/server/internal/handler/dto/sdi"
	"github.com/unicomhub/uniwork/server/internal/handler/dto/sdo"
	"github.com/unicomhub/uniwork/server/internal/middleware"
	"github.com/unicomhub/uniwork/server/internal/service"
)

func (h *handlers) listChatThreadMessages(w http.ResponseWriter, r *http.Request) {
	in := service.ListChatMessagesInput{Limit: 50}
	if v := r.URL.Query().Get("limit"); v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			in.Limit = n
		}
	}
	if v := r.URL.Query().Get("before"); v != "" {
		if t, err := time.Parse(time.RFC3339, v); err == nil {
			in.Before = &t
		}
	}
	msgs, err := h.Chat.ListThreadMessages(
		r.Context(),
		middleware.UserID(r.Context()),
		chi.URLParam(r, "workspaceID"),
		chi.URLParam(r, "roomID"),
		chi.URLParam(r, "messageID"),
		in,
	)
	if err != nil {
		h.mapServiceError(w, err)
		return
	}
	out := make([]sdo.ChatMessageDTO, 0, len(msgs))
	for _, m := range msgs {
		out = append(out, toChatMessageDTO(m))
	}
	respondJSON(w, http.StatusOK, map[string]any{"messages": out})
}

func (h *handlers) sendChatThreadMessage(w http.ResponseWriter, r *http.Request) {
	var in sdi.SendChatMessageSDI
	if !decode(w, r, &in, maxJSONBody) {
		return
	}
	msg, err := h.Chat.SendThreadReply(
		r.Context(),
		middleware.UserID(r.Context()),
		chi.URLParam(r, "workspaceID"),
		chi.URLParam(r, "roomID"),
		chi.URLParam(r, "messageID"),
		service.SendChatMessageInput{
			Body: in.Body, ReplyToMessageID: in.ReplyToMessageID,
			Priority: in.Priority, ClientMsgID: in.ClientMsgID,
		},
	)
	if err != nil {
		h.mapServiceError(w, err)
		return
	}
	respondJSON(w, http.StatusCreated, map[string]any{"message": toChatMessageDTO(msg)})
}

func (h *handlers) followChatThread(w http.ResponseWriter, r *http.Request) {
	if err := h.Chat.FollowThread(
		r.Context(),
		middleware.UserID(r.Context()),
		chi.URLParam(r, "workspaceID"),
		chi.URLParam(r, "messageID"),
	); err != nil {
		h.mapServiceError(w, err)
		return
	}
	respondJSON(w, http.StatusOK, map[string]any{"ok": true})
}

func (h *handlers) unfollowChatThread(w http.ResponseWriter, r *http.Request) {
	if err := h.Chat.UnfollowThread(
		r.Context(),
		middleware.UserID(r.Context()),
		chi.URLParam(r, "workspaceID"),
		chi.URLParam(r, "messageID"),
	); err != nil {
		h.mapServiceError(w, err)
		return
	}
	respondJSON(w, http.StatusOK, map[string]any{"ok": true})
}

func (h *handlers) markChatThreadRead(w http.ResponseWriter, r *http.Request) {
	if err := h.Chat.MarkThreadRead(
		r.Context(),
		middleware.UserID(r.Context()),
		chi.URLParam(r, "workspaceID"),
		chi.URLParam(r, "messageID"),
	); err != nil {
		h.mapServiceError(w, err)
		return
	}
	respondJSON(w, http.StatusOK, map[string]any{"ok": true})
}

func (h *handlers) listChatThreads(w http.ResponseWriter, r *http.Request) {
	unreadOnly := r.URL.Query().Get("unread") == "1"
	limit := 50
	if v := r.URL.Query().Get("limit"); v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			limit = n
		}
	}
	rows, err := h.Chat.ListFollowedThreads(
		r.Context(),
		middleware.UserID(r.Context()),
		chi.URLParam(r, "workspaceID"),
		unreadOnly,
		limit,
	)
	if err != nil {
		h.mapServiceError(w, err)
		return
	}
	out := make([]sdo.ChatThreadDTO, 0, len(rows))
	for _, t := range rows {
		dto := sdo.ChatThreadDTO{
			ThreadRootID: t.ThreadRootID,
			RoomID:       t.RoomID,
			WorkspaceID:  t.WorkspaceID,
			RootBody:     t.RootBody,
			RootSenderID: t.RootSenderID,
			ReplyCount:   t.ReplyCount,
			RootCreated:  t.RootCreated.Format(time.RFC3339),
			Unread:       t.Unread,
			Reason:       t.Reason,
		}
		if t.LastReplyAt != nil {
			dto.LastReplyAt = t.LastReplyAt.Format(time.RFC3339)
		}
		out = append(out, dto)
	}
	respondJSON(w, http.StatusOK, sdo.ChatThreadListSDO{Threads: out})
}
